src/sspmprk.jl

```diff
@@ -0,0 +1,908 @@
+    end
+
+    # solve linear system
+    linprob = LinearProblem(M, rhs)
+    sol = solve(linprob, alg.linsolve)
+    u2 = sol.u
+    u = u2
+    integrator.stats.nsolve += 1
+
+    # compute Patankar weight denominator
+    ρ = n1 * u + n2 * u .^ 2 ./ σ
+    # avoid division by zero due to zero Patankar weights
+    ρ = add_small_constant(ρ, small_constant)
+
+    P2 = f.p(u, p, t + β10 * dt)
+    Ptmp = β20 * P + β21 * P2
+    integrator.stats.nf += 1
+
+    # build linear system matrix and rhs
+    if f isa PDSFunction
+        d2 = f.d(u, p, t + β10 * dt)  # evaluate nonconservative destruction terms
+        dtmp = β20 * d + β21 * d2
+        rhs = α20 * uprev + α21 * u2 + dt * diag(Ptmp)
+        M = build_mprk_matrix(Ptmp, ρ, dt, dtmp)
+
+    else
+        rhs = α20 * uprev + α21 * u2
+        M = build_mprk_matrix(Ptmp, ρ, dt)
+    end
+
+    # solve linear system
+    linprob = LinearProblem(M, rhs)
+    sol = solve(linprob, alg.linsolve)
+    u = sol.u
+    integrator.stats.nsolve += 1
+
+    # compute Patankar weight denominator
+    σ = σ .^ (1 - s) .* u2 .^ s
+    # avoid division by zero due to zero Patankar weights
+    σ = add_small_constant(σ, small_constant)
+
+    Ptmp = η3 * P + η4 * P2
+
+    # build linear system matrix and rhs
+    if f isa PDSFunction
+        dtmp = η3 * d + η4 * d2
+
+        # see (3.25 f) in original paper
+        rhs = η1 * uprev + η2 * u2 + dt * (η5 * diag(P) + η6 * diag(P2))
+
+        M = build_mprk_matrix(Ptmp, σ, dt, dtmp)
+    else
+        rhs = η1 * uprev + η2 * u2
+        M = build_mprk_matrix(Ptmp, σ, dt)
+    end
+
+    # solve linear system
+    linprob = LinearProblem(M, rhs)
+    sol = solve(linprob, alg.linsolve)
+    σ = sol.u
+    integrator.stats.nsolve += 1
+
+    # compute Patankar weight denominator
+    σ = σ + z .* uprev .* u ./ ρ
+    # avoid division by zero due to zero Patankar weights
+    σ = add_small_constant(σ, small_constant)
+
+    P3 = f.p(u, p, t + c3 * dt)
+    Ptmp = β30 * P + β31 * P2 + β32 * P3
+    integrator.stats.nf += 1
+
+    # build linear system matrix
+    if f isa PDSFunction
+        d3 = f.d(u, p, t + c3 * dt)  # evaluate nonconservative destruction terms
+        dtmp = β30 * d + β31 * d2 + β32 * d3
+        rhs = α30 * uprev + α31 * u2 + α32 * u + dt * diag(Ptmp)
+        M = build_mprk_matrix(Ptmp, σ, dt, dtmp)
+    else
+        rhs = α30 * uprev + α31 * u2 + α32 * u
+        M = build_mprk_matrix(Ptmp, σ, dt)
+    end
+
+    # solve linear system
+    linprob = LinearProblem(M, rhs)
+    sol = solve(linprob, alg.linsolve)
+    u = sol.u
+    integrator.stats.nsolve += 1
+
+    #TODO: Figure out if a second order approximation of the solution
+    # is hidden somewhere.
+    #=
+    tmp = u - σ
+    atmp = calculate_residuals(tmp, uprev, u, integrator.opts.abstol,
+                               integrator.opts.reltol, integrator.opts.internalnorm, t)
+    integrator.EEst = integrator.opts.internalnorm(atmp, t)
+    =#
```

Comment:
The last Patankar weight denominator has to be second-order accurate, isn't it?

Reply:
Unfortunatelly that's not so clear. The schemes based on the SSP formulation are somewhat different from "regular" MPRK schemes. See Thereom 2.1 and Theorem 2.2 in https://doi.org/10.1007/s10915-018-0881-9. The order conditions shown there contain all Patankar-weights at once. 

Another difference is shown in (3.25f) of the same paper, where evaluations of the function F, which correspond to production terms which are not treated with the Patankar-Trick,  are multiplied with different weights than the other production and destruction terms. 

So I think that a closer inspection is indeed needed here.

Reply:
I see, thanks